File: frontend/src/types/workflow.ts
export interface JiraFieldSchema {
  id: string;
  name: string;
  type: "string" | "number" | "array" | "option" | "user" | "date";
  required: boolean;
  allowed_values?: string[] | null;
}

export interface JiraUser {
  account_id: string;
  display_name: string;
  email?: string;
}

export interface JiraCredentials {
  base_url: string;
  email: string;
  username: string;
  api_token: string;
  instance_type: "cloud" | "server_basic" | "server_pat";
}

export interface WorkflowStep {
  id: string;
  name: string;
  connector_type: "http" | "jenkins" | "argocd" | "slack" | "teams" | "jira" | "email" | "k8s" | "ai_analyze";
  config: Record<string, unknown>;
  on_failure: "stop" | "continue";
  output_var: string;
}

export interface Workflow {
  id: number;
  name: string;
  description: string;
  workflow_type?: string;
  trigger_type: "manual" | "webhook" | "cron" | "all" | "alert";
  webhook_token: string;
  cron_schedule: string | null;
  trigger_config?: string | null;
  input_schema: Record<string, string>;
  steps: WorkflowStep[];
  created_by: string;
  created_at: string;
  updated_at: string;
}

export interface WorkflowCreate {
  name: string;
  description?: string;
  trigger_type?: string;
  cron_schedule?: string;
  trigger_config?: string;
  input_schema?: Record<string, string>;
  steps?: WorkflowStep[];
}

export interface StepResult {
  step_id: string;
  step_name: string;
  status: "pending" | "running" | "passed" | "failed" | "skipped";
  started_at: string | null;
  completed_at: string | null;
  output: Record<string, unknown> | null;
  error: string | null;
}

export interface WorkflowRun {
  id: number;
  workflow_id: number;
  triggered_by: "manual" | "webhook" | "cron";
  trigger_input: Record<string, unknown>;
  status: "pending" | "running" | "completed" | "failed";
  started_at: string;
  completed_at: string | null;
  step_results: StepResult[];
  ai_summary: string | null;
}

export interface WorkflowSSEEvent {
  type: "step_update" | "step" | "result" | "completed" | "ping" | "error";
  step_id?: string;
  status?: string;
  message?: string;
  run_id?: number;
  error?: string;
}

File: backend/app/models/workflow.py
from typing import Optional, Any, Dict, List
from datetime import datetime, timezone
import uuid
from sqlmodel import Field, SQLModel, Column, JSON


class Workflow(SQLModel, table=True):
    __tablename__ = "workflows"

    id: Optional[int] = Field(default=None, primary_key=True)
    name: str = Field(index=True)
    description: str = Field(default="")
    trigger_type: str = Field(default="manual")  # manual|webhook|cron|all|alert
    webhook_token: str = Field(
        default_factory=lambda: str(uuid.uuid4()), unique=True, index=True
    )
    cron_schedule: Optional[str] = Field(default=None)
    trigger_config: Optional[str] = Field(default=None)  # JSON: {"alert_name": "CrashLoopBackOff"}
    input_schema: Dict[str, Any] = Field(default_factory=dict, sa_column=Column(JSON))
    steps: List[Dict[str, Any]] = Field(default_factory=list, sa_column=Column(JSON))
    created_by: str = Field(default="")
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    # Agent fields (null/empty for scripted workflows)
    workflow_type: str = Field(default="scripted")  # scripted|agent
    agent_goal: Optional[str] = Field(default=None)
    agent_approved_tools: List[Dict[str, Any]] = Field(
        default_factory=list, sa_column=Column(JSON)
    )
    agent_cluster_ids: List[int] = Field(
        default_factory=list, sa_column=Column(JSON)
    )
    agent_minion_ids: List[str] = Field(
        default_factory=list, sa_column=Column(JSON)
    )
    agent_max_retries: int = Field(default=3)
    agent_timeout_seconds: int = Field(default=900)
    agent_approval_timeout_seconds: int = Field(default=600)
    # Notification config: {"slack": {"enabled": bool, "webhook_url": str},
    #                        "teams": {"enabled": bool, "webhook_url": str},
    #                        "jira":  {"enabled": bool, "base_url": str, "project_key": str,
    #                                  "issue_type": str, "email": str, "api_token": str}}
    agent_notifications: Dict[str, Any] = Field(default_factory=dict, sa_column=Column(JSON))


class WorkflowRun(SQLModel, table=True):
    __tablename__ = "workflow_runs"

    id: Optional[int] = Field(default=None, primary_key=True)
    workflow_id: int = Field(foreign_key="workflows.id", index=True)
    triggered_by: str = Field(default="manual")  # manual|webhook|cron
    triggered_by_user_id: Optional[int] = Field(default=None)
    trigger_input: Dict[str, Any] = Field(default_factory=dict, sa_column=Column(JSON))
    status: str = Field(default="pending")  # pending|running|completed|failed|awaiting_approval
    started_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    completed_at: Optional[datetime] = Field(default=None)
    step_results: List[Dict[str, Any]] = Field(default_factory=list, sa_column=Column(JSON))
    ai_summary: Optional[str] = Field(default=None)
